File: Code/PS2/Matlab/initialize.m
function [oldvalue,oldx] = initialize(dtable,nfirms,wmax,binom,newvalue,newx)
% This procedure takes the solved newx, newvalue matrix for the nfirms - 1
% problem, and puts them into the nfirms matrices oldx, oldvalue, for use
% as starting values
oldx = zeros(wmax,nfirms);
oldvalue = zeros(wmax,nfirms);
if nfirms == 1
    oldvalue = 1+0.1*(1:wmax)'; % newvalue and newx don't exist for nfirms-1=0
else
    for w = 1:wmax
        tuple = qdecode(w,dtable);
        % Initialize by mapping the current w to the corresponding nfirms-1 
        % equilibrium
        % (ignore the last firm) 
        n = encode(tuple(1:nfirms-1),nfirms-1,binom);
        oldvalue(w,1:nfirms-1) = newvalue(n,1:nfirms-1);
        oldx(w,1:nfirms-1) = newx(n,1:nfirms-1);

        % Initialize the last firm by ignoring the second last firm, 
        % i.e., swap the last two firms and set the last firm to 0 state
        tuple(nfirms-1) = tuple(nfirms);
        tuple(nfirms) = 0;
        n = encode(tuple,nfirms,binom);
        oldvalue(w,nfirms) = oldvalue(n,nfirms-1);
        oldx(w,nfirms) =  oldx(n,nfirms-1);
    end
end
end

File: Code/PS2/Python/initialize.py
def initialize(dtable, nfirms, wmax, binom, newvalue, newx):
    """
    Initializes the value and policy functions for equilibrium computation.

    Args:
        dtable (np.ndarray): Decoding table for states.
        nfirms (int): Number of firms.
        wmax (int): Maximum number of states.
        binom (np.ndarray): Binomial coefficient matrix.
        newvalue (np.ndarray): Solved newvalue matrix for (nfirms - 1).
        newx (np.ndarray): Solved newx matrix for (nfirms - 1).

    Returns:
        tuple: (oldvalue, oldx) - Initial value function and policy function.
    """
    oldx = np.zeros((wmax, nfirms))  # Initialize policy function
    oldvalue = np.zeros((wmax, nfirms))  # Initialize value function

    if nfirms == 1:
        oldvalue = 1 + 0.1 * np.arange(1, wmax + 1)  # Base case for a single firm
    else:
        for w in range(wmax):
            tuple_w = qdecode(w, dtable)  # Decode the state

            # Map current state to corresponding (nfirms-1) equilibrium
            n = encode(tuple_w[:nfirms - 1], nfirms - 1, binom)
            oldvalue[w, :nfirms - 1] = newvalue[n, :nfirms - 1]
            oldx[w, :nfirms - 1] = newx[n, :nfirms - 1]

            # Initialize the last firm by swapping the last two firms and setting last firm to state 0
            tuple_w[nfirms - 2] = tuple_w[nfirms - 1]  # Swap last two firms
            tuple_w[nfirms - 1] = 0  # Set last firm to zero state
            n = encode(tuple_w, nfirms, binom)

            # Copy the results from (nfirms-1) game to initialize last firm
            oldvalue[w, nfirms - 1] = oldvalue[n, nfirms - 2]
            oldx[w, nfirms - 1] = oldx[n, nfirms - 2]

    return oldvalue, oldx
